### PR TITLE
 feat(uploads): resume failed uploads from uploads manager

### PR DESCRIPTION
**Pull request by rosepierce**
Adding functionality to resume uploads for files of size >100 MB that have failed from where in the upload the file failed, rather than restarting the upload from the beginning. Additionally adding a button on the Uploads Manager header which will resume any items in the Uploads Manager that have failed.

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @rosepierce has signed the CLA. Thanks for the pull request!

File: src/elements/content-uploader/UploadsManagerAction.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    onClick: Function,
+};
+
+const UploadsManagerAction = ({ onClick, numFailedUploads }: Props) => {
+    const handleResumeClick = event => {
+        event.stopPropagation();
+        onClick(STATUS_ERROR);
+    };
+
+    const resumeMessage =
```
Comment (priyajeet) [inline src/elements/content-uploader/UploadsManagerAction.js]: Similar to a prior comment:

const resumeMessage = numFailedUploads > 1 ? message.resumeAll : messages.resume;
...
<FormattedMessage {...resumeMessage} />

Comment (priyajeet) [reply src/elements/content-uploader/UploadsManagerAction.js]: And additionally, does numFailedUploads need to be passed in as-is, or can we have a boolean prop `hasMultipleUploadsFailed` which abstracts out how its calculated from the perspective of this function

File: src/api/uploads/MultiputUpload.js [hunk 2]
```diff
@@ -329,6 +341,198 @@ class MultiputUpload extends BaseMultiput {
...
+        }
+
+        if (errorData && errorData.status === 429) {
+            let retryAfterMs = DEFAULT_RETRY_DELAY_MS;
+            if (errorData.headers) {
+                const retryAfterSec = parseInt(
+                    errorData.headers['retry-after'] || errorData.headers.get('Retry-After'),
+                    10,
+                );
+                if (!Number.isNaN(retryAfterSec)) {
```
Comment (priyajeet) [inline src/api/uploads/MultiputUpload.js]: FYI, Number.isNaN doesn't work in IE. Likely works fine in webapp due to polyfill.
But can use `import isNaN from 'lodash/isNaN';` which we do in other files.

File: src/api/uploads/MultiputUpload.js [hunk 3]
```diff
@@ -329,6 +341,198 @@ class MultiputUpload extends BaseMultiput {
...
+        if (this.numResumeRetries > this.config.retries) {
+            this.errorCallback(errorData);
+            return;
+        }
+
+        if (errorData && errorData.status === 429) {
+            let retryAfterMs = DEFAULT_RETRY_DELAY_MS;
+            if (errorData.headers) {
+                const retryAfterSec = parseInt(
+                    errorData.headers['retry-after'] || errorData.headers.get('Retry-After'),
```
Comment (priyajeet) [inline src/api/uploads/MultiputUpload.js]: I think this code is probably repeated from somewhere in the inheritance, can make it a function and just call that to get retury after

File: src/api/uploads/MultiputUpload.js [hunk 4]
```diff
@@ -329,6 +341,198 @@ class MultiputUpload extends BaseMultiput {
...
+            // Restart upload process for errors resulting from invalid session
+            this.parts.forEach(part => {
+                part.cancel();
+            });
+            this.parts = [];
+            // Reset information about uploaded parts
+            this.fileSha1 = null;
+            this.totalUploadedBytes = 0;
+            this.numPartsNotStarted = 0;
+            this.numPartsDigestComputing = 0;
```
Comment (priyajeet) [inline src/api/uploads/MultiputUpload.js]: Isn't some/most of this also happening in the constructor? Maybe should be extracted out as a reset method?

Likewise for anything else thats being duplicated in bulk

File: src/elements/content-uploader/ContentUploader.js [hunk 5]
```diff
@@ -746,7 +786,7 @@ class ContentUploader extends Component<Props, State> {
         const { onUpload, useUploadsManager } = this.props;
 
         item.progress = 100;
-        if (!item.error) {
+        if (item.status !== STATUS_ERROR) {
```
Comment (danubevictoria) [inline src/elements/content-uploader/ContentUploader.js]: Why are we changing this to check status? Should it be that we only mark the status as complete if there is no error?

File: src/elements/content-uploader/ContentUploader.js [hunk 6]
```diff
@@ -917,14 +958,36 @@ class ContentUploader extends Component<Props, State> {
                 this.removeFileFromUploadQueue(item);
                 break;
             case STATUS_ERROR:
-                this.resetFile(item);
-                this.uploadFile(item);
+                if (isResumableUploadsEnabled && chunked && item.api.sessionId) {
```
Comment (danubevictoria) [inline src/elements/content-uploader/ContentUploader.js]: Are there any clients where chunked upload is not supported?

File: src/api/uploads/__tests__/MultiputUpload-test.js [hunk 7]
```diff
@@ -418,6 +419,183 @@ describe('api/uploads/MultiputUpload', () => {
         });
     });
 
+    describe('getSessionSuccessHandler()', () => {
+        test('should commit if all parts have been uploaded', () => {
+            // Setup
+            multiputUploadTest.numPartsUploaded = 20;
+            const response = {
```
Comment (mickr) [inline src/api/uploads/__tests__/MultiputUpload-test.js]: Can this be moved to a `beforeEach`?

File: src/api/uploads/MultiputUpload.js [hunk 8]
```diff
@@ -340,7 +523,9 @@ class MultiputUpload extends BaseMultiput {
      * @return {Promise}
      */
     async sessionErrorHandler(error: ?Error, logEventType: string, logMessage?: string): Promise<any> {
-        this.destroy();
+        if (!this.isResumableUploadsEnabled) {
```
[the remark this answers is not in the document]
Comment (rosepierce) [reply src/api/uploads/MultiputUpload.js]: MultiputUpload is only used for the upload process after a check for both `chunked` and if multiput is supported in ContentUploader, and if these aren't supported for the given item and client then the upload process goes through PlainUpload instead, so I don't think those checks need to be performed here or at line 546.

File: src/api/uploads/MultiputUpload.js [hunk 9]
```diff
@@ -329,6 +341,198 @@ class MultiputUpload extends BaseMultiput {
...
+    }
+
+    /**
+     * Handle error from getting upload session.
+     * Restart uploads without valid sessions from the beginning of the upload process.
+     *
+     * @param error
+     * @return {void}
+     */
+    getSessionErrorHandler(error: Object): void {
```
Comment (mickr) [inline src/api/uploads/MultiputUpload.js]: Is there a more descriptive type that can be used for the error object?

File: src/api/uploads/MultiputUpload.js [hunk 10]
```diff
@@ -329,6 +381,171 @@ class MultiputUpload extends BaseMultiput {
...
+
+    /**
+     * Get session information from API.
+     * Uses session info to commit a complete session or continue an in-progress session.
+     *
+     * @private
+     * @return {void}
+     */
+    getSessionInfo = async (): Promise<any> => {
+        const uploadUrl = this.getBaseUploadUrl();
```
Comment (Alhalalsheh) [inline src/api/uploads/MultiputUpload.js]: would the uploadUrl include the /api/2.0 or 2.1?

File: src/api/uploads/MultiputUpload.js [hunk 11]
```diff
@@ -329,6 +381,171 @@ class MultiputUpload extends BaseMultiput {
...
+        const sessionUrl = `${uploadUrl}/files/upload_sessions/${this.sessionId}`;
```
Comment (Alhalalsheh) [inline src/api/uploads/MultiputUpload.js]: make sure this is a GET ... just confirming

File: src/api/uploads/MultiputUpload.js [hunk 12]
```diff
@@ -329,6 +381,171 @@ class MultiputUpload extends BaseMultiput {
...
+
+    /**
+     * Handles a getSessionInfo success and either commits the session or continues to process
+     * the parts that still need to be uploaded.
+     *
+     * @param response
+     * @return {void}
+     */
+    getSessionSuccessHandler(data: any): void {
+        const { total_parts, part_size, session_endpoints } = data;
```
Comment (Alhalalsheh) [inline src/api/uploads/MultiputUpload.js]: it worth checking "session_expires_at". if the session is already expired, you just can not resume the download

File: src/api/uploads/MultiputUpload.js [hunk 13]
```diff
@@ -329,6 +381,171 @@ class MultiputUpload extends BaseMultiput {
...
+                this.numPartsNotStarted += 1;
+            }
+            nextUploadIndex += 1;
+        }
+
+        if (this.numPartsUploaded === total_parts) {
+            // Commit sessions that have all the parts of the file uploaded
+            this.commitSession();
+        } else {
+            this.processNextParts();
```
Comment (Alhalalsheh) [inline src/api/uploads/MultiputUpload.js]: im assuming processNextParts will eventually direct you to the commitSession flow ... and im assuming commitSession will perform that sha calculation to insure file integrity

File: src/api/uploads/MultiputUpload.js (hunk 10 above)
Comment (rosepierce) [reply src/api/uploads/MultiputUpload.js]: This uploadUrl uses /api/2.0, which is also being used for the rest of the upload process - is 2.0 or 2.1 what should be used to use chunked uploads?

File: src/api/uploads/MultiputUpload.js (hunk 13 above)
Comment (rosepierce) [reply src/api/uploads/MultiputUpload.js]: yes, when all the parts have been successfully uploaded processNextParts calls commitSession, which sends the sha calculation that has been performed during the file upload with the POST to commit the session

File: src/api/uploads/MultiputUpload.js (hunk 12 above)
Comment (rosepierce) [reply src/api/uploads/MultiputUpload.js]: From what I saw in the documentation for the chunked upload API, I thought that any request made with a session ID for a session that has expired would return a error, and so the scenario where the session has expired is being handled in the getSessionErrorHandler. Is this not necessarily the case that an error would be returned?

File: src/api/uploads/MultiputUpload.js (hunk 10 above)
Comment (Alhalalsheh) [reply src/api/uploads/MultiputUpload.js]: i see the browser today is also using 2.0, while we recommend using the latest version, i have no strong opinion to push you towards 2.1. 2.0 is fine by me

File: src/api/uploads/MultiputUpload.js (hunk 13 above)
Comment (Alhalalsheh) [reply src/api/uploads/MultiputUpload.js]: cool ... do we have a unit test for this? sorry i did not go through unit tests ... but just wanted to confirm we have a unit test to exercise/confirm sha calculation at client side on multiput (chunked) upload scenario

File: src/api/uploads/MultiputUpload.js (hunk 12 above)
Comment (Alhalalsheh) [reply src/api/uploads/MultiputUpload.js]: if you know that your session is expired why not to optimize and fail (or restart the upload) immediately? yet if you feel adding an extra check for session_expires_at will hurt the code readability i would skip it!

it is up to you to decide, yet if you decided to add session_expires_at check, make sure NOT to remove the "scenario where the session has expired is being handled in the getSessionErrorHandler"

**Pull request closed without merging**